### PR TITLE
L2 Genesis Predeploy Updates

### PR DESCRIPTION
**Pull request by Spacesai1or**
This PR is adding the predeploys at the address these contracts are deployed to on OP mainnet (some of the addresses differ from deployments on other L2s and ETH mainnet):

- `Safe`
- `SafeL2`
- `MultiSend`
- `MultiSendCallOnly`
- `MultiCall3`
- `Create2Deployer`
- `Permit2`
- EIP-4337's `EntryPoint`
- `SafeSingletonFactory`
- `DeterministicDeploymentProxy`

The deployed bytecode used for the Genesis `alloc` for each contract was mostly generated by BindGen (#7762), with the exceptions of: `MultiSend`, `Permit2`, and `EntryPoint`. These contracts contained `immutable` variables that were dependent on `block.chainid` and the contract's address. They were deployed to a simulated backend where the chain ID was set to `config.L2ChainID`, using Arachnid's proxy deployer to resolve their expected addresses, and their deployed bytecode was then retrieved

### CODE REVIEW

**Comment by Spacesai1or** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #7762** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/7762?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #7804** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/7804?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
      * **PR #8071** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8071?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/7804?utm_source=stack-comment).

File: op-chain-ops/cmd/check-l2/main.go [hunk 1]
```diff
@@ -264,6 +264,41 @@ func checkPredeployConfig(client *ethclient.Client, name string) error {
 			if err := checkEAS(p, client); err != nil {
 				return err
 			}
+
+		case predeploys.SafeL2Addr:
+			if err := checkPredeployBytecode(p, "SafeL2", client, []byte(bindings.SafeL2ABI)); err != nil {
+				return err
+			}
+
+		case predeploys.MultiSendAddr:
```
Comment (tynes) [inline op-chain-ops/cmd/check-l2/main.go]: You can see in the autogenerated code that `xxxABI` is deprecated in favor of `xxxMetaData.ABI`

```

// MultiSendABI is the input ABI used to generate the binding from.
// Deprecated: Use MultiSendMetaData.ABI instead.
```

File: op-chain-ops/deployer/deployer.go [hunk 2]
```diff
@@ -42,19 +44,38 @@ type Deployer func(*backends.SimulatedBackend, *bind.TransactOpts, Constructor)
...
+
+	return backend, nil
 }
 
 // NewL2Backend returns a SimulatedBackend suitable for L2.
 // It has the latest L2 hardforks enabled.
-func NewL2Backend() *backends.SimulatedBackend {
-	return NewBackendWithGenesisTimestamp(0, false)
+func NewL2Backend() (*backends.SimulatedBackend, error) {
+	backend, err := NewBackendWithGenesisTimestamp(ChainID, 0, true, make(map[string]*common.Address))
```
Comment (tynes) [inline op-chain-ops/deployer/deployer.go]: I noticed that `true` is passed now instead of `false`, why did you change that? iirc `true` will activate hardforks that are not yet available on L2

File: op-chain-ops/deployer/deployer.go [hunk 3]
```diff
@@ -145,6 +179,70 @@ func Deploy(backend *backends.SimulatedBackend, constructors []Constructor, cb D
...
+	}
+
+	return code, nil
+}
+
 func u64ptr(n uint64) *uint64 {
 	return &n
 }
+
+func create2Address(creatorAddress, salt, initCode []byte) []byte {
```
Comment (tynes) [inline op-chain-ops/deployer/deployer.go]: Can you write unit tests for this?

File: op-chain-ops/deployer/deployer.go [hunk 4]
```diff
@@ -42,19 +44,38 @@ type Deployer func(*backends.SimulatedBackend, *bind.TransactOpts, Constructor)
 
 // NewL1Backend returns a SimulatedBackend suitable for L1. It has
 // the latest L1 hardforks enabled.
-func NewL1Backend() *backends.SimulatedBackend {
-	return NewBackendWithGenesisTimestamp(0, true)
+func NewL1Backend() (*backends.SimulatedBackend, error) {
+	backend, err := NewBackendWithGenesisTimestamp(ChainID, 0, true, make(map[string]*common.Address))
```
Comment (tynes) [inline op-chain-ops/deployer/deployer.go]: can you pass `nil` instead of `make(map...)`?

File: op-chain-ops/genesis/layer_two.go [hunk 5]
```diff
@@ -51,27 +55,70 @@ func BuildL2Genesis(config *DeployConfig, l1StartBlock *types.Block) (*core.Gene
...
-			// there is no governance token configured, so skip the governance token predeploy
-			log.Warn("Governance is not enabled, skipping governance token predeploy.")
-			continue
-		}
+
 		codeAddr := addr
-		if predeploys.IsProxied(addr) {
-			codeAddr, err = AddressToCodeNamespace(addr)
+		switch name {
+		case "SafeL2":
```
Comment (tynes) [inline op-chain-ops/genesis/layer_two.go]: you can do multiple matches with a single case with commas in between the matches

File: op-chain-ops/deployer/deployer.go (hunk 2 above)
Comment (Spacesai1or) [reply op-chain-ops/deployer/deployer.go]: Copy/paste error, thank you!

File: op-chain-ops/deployer/deployer.go [hunk 6]
```diff
@@ -42,19 +44,38 @@ type Deployer func(*backends.SimulatedBackend, *bind.TransactOpts, Constructor)
 
 // NewL1Backend returns a SimulatedBackend suitable for L1. It has
 // the latest L1 hardforks enabled.
-func NewL1Backend() *backends.SimulatedBackend {
-	return NewBackendWithGenesisTimestamp(0, true)
+func NewL1Backend() (*backends.SimulatedBackend, error) {
+	backend, err := NewBackendWithGenesisTimestamp(ChainID, 0, true, nil)
```
Comment (tynes) [inline op-chain-ops/deployer/deployer.go:48]: There is no real need to check for an error here - notice its just being passed thru. It would only be useful to do this if you annotated the error like so `return nil, fmt.Errorf("cannot create L1 backend: %w, err")`

File: op-chain-ops/deployer/deployer.go [hunk 7]
```diff
@@ -90,13 +126,11 @@ func NewBackendWithGenesisTimestamp(ts uint64, shanghai bool) *backends.Simulate
 			Config:     &chainConfig,
 			Timestamp:  ts,
 			Difficulty: big.NewInt(0),
-			Alloc: core.GenesisAlloc{
-				crypto.PubkeyToAddress(TestKey.PublicKey): {Balance: thousandETH},
-			},
-			GasLimit: 15000000,
+			Alloc:      alloc,
+			GasLimit:   15000000,
```
Comment (tynes) [inline op-chain-ops/deployer/deployer.go]: Can you update the gaslimit to 30 million?

File: op-chain-ops/deployer/deployer.go (hunk 3 above)
Comment (tynes) [inline op-chain-ops/deployer/deployer.go]: should this return a `common.Address` rather than `[]byte`?
